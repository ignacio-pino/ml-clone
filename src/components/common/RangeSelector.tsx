import { ArrowForwardIcon } from "@chakra-ui/icons";
import { Divider, HStack, IconButton, Input } from "@chakra-ui/react";
import { useState } from "react";
import { useLocation, useNavigate, createSearchParams } from "react-router-dom";

export default function RangeSelector() {
  const [values, setValues] = useState({ minValue: "", maxValue: "" });

  const { search } = useLocation();
  const searchParams = new URLSearchParams(search);

  const navigate = useNavigate();

  let params = Object.fromEntries(searchParams.entries());

  const handleClick = () => {
    params = {
      ...params,
      price: values.minValue
        ? values.maxValue
          ? `${values.minValue}.0-${values.maxValue}.0`
          : `${values.minValue}.0-*`
        : `*-${values.maxValue}.0`,
    };
    navigate({
      pathname: "/products",
      search: `${createSearchParams(params)}`,
    });
    setValues({ minValue: "", maxValue: "" });
  };

  return (
    <HStack>
      <Input
        type="number"
        placeholder="Min."
        variant={"outline"}
        h={"2rem"}
        w={"5rem"}
        value={values.minValue}
        onChange={(event) => {
          setValues({
            ...values,
            minValue: event.currentTarget.value,
          });
        }}
      />
      <Divider orientation="horizontal" w={3} />
      <Input
        type="number"
        placeholder="Max."
        variant={"outline"}
        h={"2rem"}
        w={"5rem"}
        value={values.maxValue}
        onChange={(event) => {
          setValues({
            ...values,
            maxValue: event.currentTarget.value,
          });
        }}
      />
      <IconButton
        aria-label="Buscar rango"
        icon={<ArrowForwardIcon />}
        variant={"outline"}
        borderRadius={"10px"}
        size={"sm"}
        onClick={() => handleClick()}
      />
    </HStack>
  );
}
