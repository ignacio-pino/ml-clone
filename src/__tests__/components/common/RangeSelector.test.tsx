import { render, screen, fireEvent } from "@testing-library/react";
import RangeSelector from "../../../components/common/RangeSelector";
import { MemoryRouter, createSearchParams } from "react-router-dom";

const mockedUseNavigate = vitest.fn();

vitest.mock("react-router-dom", () => ({
  ...require("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("RangeSelector Component", () => {
  beforeEach(() => {
    mockedUseNavigate.mockClear();
  });

  it("matches snapshot", () => {
    const params = { category: "test" };
    const container = render(
      <MemoryRouter
        initialEntries={[
          { pathname: "/products", search: `${createSearchParams(params)}` },
        ]}
      >
        <RangeSelector />
      </MemoryRouter>
    ).container;

    expect(container).toMatchSnapshot();
  });

  it("updates input values correctly", () => {
    render(
      <MemoryRouter>
        <RangeSelector />
      </MemoryRouter>
    );

    const minInput = screen.getByPlaceholderText("Min.");
    const maxInput = screen.getByPlaceholderText("Max.");

    fireEvent.change(minInput, { target: { value: "10" } });
    fireEvent.change(maxInput, { target: { value: "100" } });

    expect(minInput).toHaveDisplayValue("10");
    expect(maxInput).toHaveDisplayValue("100");
  });

  it("calls the handleClick function with correct search params", () => {
    render(
      <MemoryRouter>
        <RangeSelector />
      </MemoryRouter>
    );

    const minInput = screen.getByPlaceholderText("Min.");
    const maxInput = screen.getByPlaceholderText("Max.");
    const submitButton = screen.getByLabelText("Buscar rango");

    fireEvent.change(minInput, { target: { value: "10" } });
    fireEvent.change(maxInput, { target: { value: "100" } });

    fireEvent.click(submitButton);

    expect(mockedUseNavigate).toHaveBeenCalledWith({
      pathname: "/products",
      search: "price=10.0-100.0",
    });
  });

  it("handles only min value correctly in handleClick", () => {
    render(
      <MemoryRouter>
        <RangeSelector />
      </MemoryRouter>
    );

    const minInput = screen.getByPlaceholderText("Min.");
    const submitButton = screen.getByLabelText("Buscar rango");

    fireEvent.change(minInput, { target: { value: "10" } });

    fireEvent.click(submitButton);

    expect(mockedUseNavigate).toHaveBeenCalledWith({
      pathname: "/products",
      search: "price=10.0-*",
    });
  });

  it("handles only max value correctly in handleClick", () => {
    render(
      <MemoryRouter>
        <RangeSelector />
      </MemoryRouter>
    );

    const maxInput = screen.getByPlaceholderText("Max.");
    const submitButton = screen.getByLabelText("Buscar rango");

    fireEvent.change(maxInput, { target: { value: "100" } });

    fireEvent.click(submitButton);

    expect(mockedUseNavigate).toHaveBeenCalledWith({
      pathname: "/products",
      search: "price=*-100.0",
    });
  });
});
