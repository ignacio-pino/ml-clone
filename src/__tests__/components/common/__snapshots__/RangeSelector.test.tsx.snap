// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RangeSelector Component > matches snapshot 1`] = `
<div>
  <div
    class="chakra-stack css-1igwmid"
  >
    <input
      class="chakra-input css-uo3m1g"
      placeholder="Min."
      type="number"
      value=""
    />
    <hr
      aria-orientation="horizontal"
      class="chakra-divider css-tztl8c"
    />
    <input
      class="chakra-input css-uo3m1g"
      placeholder="Max."
      type="number"
      value=""
    />
    <button
      aria-label="Buscar rango"
      class="chakra-button css-mn3g0g"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
